src/test/regress/expected/multi_subquery.out

```diff
@@ -377,7 +378,8 @@ JOIN
 ON (l_orderkey = o_orderkey + 1)
 ORDER BY l_orderkey DESC
 LIMIT 10;
-ERROR:  complex joins are only supported when all distributed tables are joined on their distribution columns with equal operator
+ERROR:  cannot perform distributed planning on this query
```

Comment:
I wonder if we could improve this error message, at least for this particular case. We've previously gotten a lot of feedback from users to provide better error messages in unsupported cases.

Reply:
Yeah, I am not happy about these errors and have looked around how to make these errors better.

Biggest problem is that we changed what we do recognised as a join. Before the `(l_orderkey = o_orderkey + 1)` clause was not recognised as a join, but now that we do not require `Var`'s to be directly on either side the clause is recognised as a join.

This causes `MultiNodeTree(...)` to be called which _always_ returns a plan, as everything will fall back to a cartesian product for joins that do not satisfy any other join order planner. The 'correct' approach would be to error in the `DualPartitionJoin` as this should be handled in the end by that planner. However it is not expected in the join order planner to error.

I wonder if removing the cartesian product planner and providing the `complex joins are only supported ....` error if the join order planner can't find anything without it. But that will change the output of a dozen or so test suites as the Cartesian product error is throughout our suite :(.

Any ideas how to provide a correct error here?

Reply:
I have cooked up something that works: https://github.com/citusdata/citus/pull/3180/commits/1e084ea85b0c2f61500a536defcdf441f1f3e9d7

Instead of always falling back to a cartesian product it will only apple the cartesian product if there are no applicable join clauses between tables. This will prevent queries like above to be planned with a cartesian product, but instead fail to plan.

For this to work the join order planner is now allowed to come up with a plan for a certain permutation. This cascades down the call chain and will give the complex query error if there is no candidate plan found to pick a best plan from. This seems to be a good tradeoff as this error reverted back while other cartesian errors remained.